cpp/src/io/parquet/reader_impl.cu

```diff
@@ -354,6 +354,44 @@ struct metadata : public FileMetaData {
   }
 };
 
+
+void reader::impl::read_column_chunks(
+    std::vector<rmm::device_buffer>& page_data,
+    hostdevice_vector<gpu::ColumnChunkDesc>& chunks,
+    size_t begin_chunk, size_t end_chunk,
+    const std::vector<size_t>& column_chunk_offsets,
+    cudaStream_t stream) {
+  // Transfer chunk data, coalescing adjacent chunks
+  for (size_t chunk = begin_chunk; chunk < end_chunk; ) {
+    const size_t io_offset = column_chunk_offsets[chunk];
+    size_t io_size = chunks[chunk].compressed_size;
+    size_t next_chunk = chunk + 1;
+    const bool is_compressed = (chunks[chunk].codec != parquet::Compression::UNCOMPRESSED);
+    while (next_chunk < end_chunk) {
+      const size_t next_offset = column_chunk_offsets[next_chunk];
+      const bool is_next_compressed = (chunks[next_chunk].codec != parquet::Compression::UNCOMPRESSED);
+      if (next_offset != io_offset + io_size || is_next_compressed != is_compressed) {
+        // Can't merge if not contiguous or mixing compressed and uncompressed
+        // Not coalescing uncompressed with compressed chunks is so that compressed buffers can be
+        // freed earlier (immediately after decompression stage) to limit peak memory requirements
+        break;
+      }
+      io_size += chunks[next_chunk].compressed_size;
+      next_chunk++;
+    }
+    uint8_t *d_compdata = nullptr;
+    if (io_size != 0) {
+      auto buffer = _source->get_buffer(io_offset, io_size);
+      page_data[chunk] = rmm::device_buffer(buffer->data(), buffer->size(), stream);
+      d_compdata = reinterpret_cast<uint8_t *>(page_data[chunk].data());
+    }
+    do {
```

Comment:
I added the check to avoid calling rmm::device_buffer::data() if io_size was null, but otherwise, io_size = 0 implies that compressed_size = 0 for all chunks, so it will correctly add zero to a uint8_t pointer, which afaik is not undefined and the output remains nullptr (note that io_size=0 is actually not expected to ever happen, implies a corrupted file, so we could add a CUDF_EXPECTS(io_size != 0)).

Reply:
Adding `CUDF_EXPECTS(io_size != 0))` fully addresses my concerns, we are definitely out of the UB-land. 
Also, `if (io_size != 0)` condition can be removed in this case.

Reply:
I suppose it might be possible that a writer could construct a rowgroup with zero rows, which could possibly still end up here (though this should normally only occur for empty df).
Perhaps one way is to just bypass the whole loop for io_size=0 and just set chunk=next_chunk (also avoids the need to initialize d_compdata=nullptr) -> I think I'll go with that.

Reply:
Updated -> can you check again ?